### ISSUE TITLE
npmInstall adds many more dependencies than package.json

### ISSUE THREAD

**Issue by chefhoobajoob** [srs/gradle-node-plugin#87]
When I run `npm install` directly from the command line in the project folder where my package.json is located, the only thing that shows up in node_modules is exactly the list of dependencies I've placed in package.json.

When I run `gradle npmInstall` using this plugin, a whole bunch of modules end up in that directory, none of which I need, and which make it impossible for me to create a distribution that includes only the dependencies I need without having to explicitly list them  all and keep them up to date in gradle.build.

Why are all of these unrelated modules placed into node_modules? If they are used by the plugin, can't I get the plugin to store the ones it needs somewhere else, and keep only the dependencies I explicitly list in package.json in node_modules? If this is possible, how do I do it?

**Comment by srs** [comment]
Hi.

Not sure on why there's things that is instaled that's not in ``package.json``. The only think the Gradle plugin does is to actually execute ``npm install``. Will check to see if I can get the same behaviour locally here. Could you post the ``package.json`` file?

**Comment by srs** [comment]
Also, please post your ``node`` configuration that you have in ``build.gradle``. Thanks.

**Comment by chefhoobajoob** [comment]
I was trying to get everything together for this issue after backing out the plugin due to the problem, and now I can't get the `npmInstall` task to run successfully.

I have node version 4.2.2 and npm version 3.4.1 installed on my system, and their bin folders are on my path environment.

I started by removing the `node_modules` folder from the project directory, adding back in the `apply plugin`, and the `node` configuration in `build.gradle`, then ran `./gradlew npmInstall`.

I created a gist with [the relevant setup here](https://gist.github.com/chefhoobajoob/16f83e57cdf4445fb244) .

Below are some quick links to the relevant pieces. 

Is there something else I need to do before running the `npmInstall` task?

[package.json](https://gist.github.com/chefhoobajoob/16f83e57cdf4445fb244#file-package-json)
[gradle apply plugin snippet](https://gist.github.com/chefhoobajoob/16f83e57cdf4445fb244#file-gradle-apply-plugin-groovy)
[gradle node config snippet](https://gist.github.com/chefhoobajoob/16f83e57cdf4445fb244#file-gradle-node-config-groovy)
[gradle npmInstall error](https://gist.github.com/chefhoobajoob/16f83e57cdf4445fb244#file-npminstall-error-txt)

**Comment by chefhoobajoob** [comment]
It seems to be complaining that it can't find `node_modules/bin/npm-cli.js`. Why is it looking there for npm instead of using my installed version?

**Comment by srs** [comment]
Thanks.

Since you have ``download`` set to ``false`` it will try to use locally installed node. If you set ``download`` to ``true`` it will try to download node specified in the ``nodeVersion`` property. 

When specifying ``npmVersion`` in the ``build.gradle`` file you instruct the plugin to actually try to install ``npm`` locally instead of using the global installed one. It could be something here that's messing things up. Try to either set ``download`` to ``true`` or remove ``npmVersion`` property from your build file to see if that helps.

I must admit that the configuration right now is a little bit messy and I will try to clean this up in the comming versions.

**Comment by chefhoobajoob** [comment]
OK - so, let's say I wanted it to work on any target system, whether they've downloaded `node` or not. If I understand correctly, the right thing to do is leave the config as is, but set `download=true`. Is that correct?

If I set `download=true`, where does `node` get downloaded to?

If the plugin downloads the `node` installation into the project's `node_modules` folder, that's what I'm trying to avoid. The only thing that should land in `node_modules`, at least in my case, is what is specified in `package.json`. Versions of `node` and `npm` that are downloaded by the plugin (I would expect) should land in some completely different location.

**Comment by chefhoobajoob** [comment]
I think maybe what I was seeing when I originally reported this issue, is perhaps the result of setting `download=true`, which I probably did at some point.

If the plugin can be changed or configured to use a directory other than `node_modules` for storing downloaded versions of `node`, `npm`, and their dependencies, then I think we'd be good to go.

**Comment by chefhoobajoob** [comment]
I can confirm that commenting out or removing `version`, `npmVersion`, and `distBaseUrl`, and leaving `download=false` still results in a failure because the plugin can't find `npm-cli.js` in the `node_modules\bin` directory.

I'm not sure why it's looking there at all. From the same command line prompt where the gradle build is failing, I can issue both `npm` and `node` commands, and the shell finds them no problem:

```
PS C:\Users\hoobajoob\Documents\Eclipse\workspace-sts-luna\project-x> node --version
v4.2.2
PS C:\Users\hoobajoob\Documents\Eclipse\workspace-sts-luna\project-x> npm --version
3.4.1
```

**Comment by srs** [comment]
Yes, when I set up projects I always set ``download=true`` so that the project build would work for everyone in the team. ``node`` binaries are downloaded into a seperate folder outside ``node_modules`` that can be set using ``workDir`` settings (which is set to ``<GRADLE_USER_HOME>/nodejs`` as default).

If you set ``npmVersion`` property then it will try to install ``npm`` (locally) using the provided ``npm`` version. This will probably pollute ``node_modules`` directory with unwanted ``npm`` dependencies. So, if you do not care about which ``npm`` version to use, then just remove ``npmVersion`` property setting.

The last part here is going to be fixed in the next or the version after that. My goal is that nothing should be put into ``node_modules`` without being in ``package.json`` file.

**Comment by srs** [comment]
The problem that you are seeing with ``download=false`` is probably a bug. It should not look into ``node_modules`` in that case. Will try to create some test cases around this problem.

**Comment by chefhoobajoob** [comment]
ok - I will try removing the version information and setting `download=true`.

**Comment by chefhoobajoob** [comment]
I can confirm that removing the version information for both `node` and `npm` does not result in anything other than the dependencies specified in `package.json` landing in `node_modules`.

And, specifying `npmVersion` is what is contaminating the project's `node_modules` folder.

Is there already a fix or version in the works to address this? Should we keep this issue open?

**Comment by chefhoobajoob** [comment]
I think I'll be able to get by just fine in the meantime, since presumably if I specify the right version of `node` I'll presumably get a version of `npm` that I can use.

**Comment by srs** [comment]
The "npmVersion" fix is not done yet. So, will keep the issue.

**Comment by chefhoobajoob** [comment]
Ok - thanks so much for your help on this.

**Comment by srs** [comment]
No problem.

**Comment by deepu105** [comment]
We are facing the same issue at JHipster as well. When download=false shouldn't it be looking for the global npm installed on the system before checking at the local node_modules folder? cc @atomfrede

**Comment by deepu105** [comment]
@srs any timeline on when this will be fixed?

**Comment by denisazevedo** [comment]
Same for me as reported by @deepu105
